functions/go/search-replace/api/search_replace.go

```diff
@@ -346,3 +350,61 @@ func (sr *SearchReplace) resultsString() string {
 	out += fmt.Sprintf("%s %d field(s)\n", action, sr.Count)
 	return out
 }
+
+// Decode decodes the input yaml RNode into SearchReplace struct
+// returns error if input yaml RNode contains invalid matcher name inputs
+func Decode(rn *yaml.RNode, fcd *SearchReplace) error {
+	dm := rn.GetDataMap()
+	if err := validateMatcherNames(dm); err != nil {
+		return err
+	}
+	fcd.ByPath = getValue(dm, ByPath)
+	fcd.ByValue = getValue(dm, ByValue)
+	fcd.ByValueRegex = getValue(dm, ByValueRegex)
+	fcd.PutValue = getValue(dm, PutValue)
+	fcd.PutComment = getValue(dm, PutComment)
+	return nil
+}
+
+// validateMatcherNames validates the input matcher names
+func validateMatcherNames(m map[string]string) error {
+	matcherSet := sets.String{}
+	matcherSet.Insert(matchers...)
+	for key := range m {
+		if !matcherSet.Has(key) {
+			return errors.Errorf("invalid matcher %q, must be one of %q", key, matchers)
+		}
+	}
+	return nil
+}
+
+// getValue returns the value for 'key' in map 'm'
+// returns empty string if 'key' doesn't exist in 'm'
+func getValue(m map[string]string, key string) string {
+	if val, ok := m[key]; ok {
+		return val
+	}
+	return ""
+}
+
+// validateMatchers validates the input matchers in SearchReplace struct
+func (sr *SearchReplace) validateMatchers() error {
+	if sr.ByValue == "" && sr.ByValueRegex == "" && sr.ByPath == "" {
+		return errors.Errorf(`at least one of [%q, %q, %q] must be provided`, ByValue, ByValueRegex, ByPath)
+	}
+
+	if sr.ByValue != "" && sr.ByValueRegex != "" {
+		return errors.Errorf(`only one of [%q, %q] can be provided`, ByValue, ByValueRegex)
+	}
+	return nil
+}
+
+const (
+	ByValue      = "by-value"
+	ByValueRegex = "by-value-regex"
+	ByPath       = "by-path"
+	PutValue     = "put-value"
+	PutComment   = "put-comment"
+)
+
+var matchers = []string{ByValue, ByValueRegex, ByPath, PutValue, PutComment}
```

Comment:
Can we avoid using package level variable?

Reply:
Converted to func.